### ISSUE TITLE
[pimatic-homeduino] weather sensors do not update

### ISSUE THREAD

**Issue by elektronaut69** [pimatic/pimatic#551]
I just flashed my nano with commit d14de3a73db28b8e91abbfbd1acde0cfd69dba99 because I had a problem with disconnects of the nano after some hours. I think the commit I used before was c7cbfdbd3dd447b3dc5bd0ed62ba14114bc0381a but I am not shure about this. 
Now my weather sensors do not update anymore. Pimatic is receiving something but seems to fail decoding the received data. 
```
debug [pimatic-homeduino]: received: [ 701, 1846, 3746, 7588 ] 01010202020102010201010101020102010201020101010202010102010101010201010101010202010103
12:37:19debug [pimatic-homeduino]: data: "RF receive 701 1846 3746 7588 0 0 0 0 01010202020102010201010101020102010201020101010202010102010101010201010101010202010103"
12:37:18debug [pimatic-homeduino]: received: [ 676, 1869, 3773, 7616 ] 01010202020102010201010101020102010201020101010202010102010101010201010101010202010103
12:37:18debug [pimatic-homeduino]: data: "RF receive 676 1869 3773 7616 0 0 0 0 01010202020102010201010101020102010201020101010202010102010101010201010101010202010103"
12:37:01debug [pimatic-homeduino]: received: [ 702, 1846, 3760, 7596 ] 01010101010102020102010101010202020101010102010202010201010101010201010201010102020103
12:37:01debug [pimatic-homeduino]: data: "RF receive 702 1846 3760 7596 0 0 0 0 01010101010102020102010101010202020101010102010202010201010101010201010201010102020103"
```

**Comment by sweetpi** [comment]
What protocol did it use?

**Comment by elektronaut69** [comment]
Weather1 and Weather3. After a cold reboot pimatic updates the weather1 sensor but not the weather 3 ones. 
```
debug [pimatic-homeduino]: received: [ 703, 1844, 3754, 7596 ] 01010101010102020102010101010202020201010102010202010201010201010201010201010102010203
12:59:32debug [pimatic-homeduino]: data: "RF receive 703 1844 3754 7596 0 0 0 0 01010101010102020102010101010202020201010102010202010201010201010201010201010102010203"
12:59:31debug [pimatic-homeduino]: received: [ 690, 1857, 3763, 7584 ] 01010101010102020102010101010202020201010102010202010201010201010201010201010102010203
12:59:31debug [pimatic-homeduino]: data: "RF receive 690 1857 3763 7584 0 0 0 0 01010101010102020102010101010202020201010102010202010201010201010201010201010102010203"
12:59:28debug [pimatic-homeduino]: weather5: { id: 154, battery: 'Good', temperature: -153.5, humidity: 47 }
12:59:28debug [pimatic-homeduino]: weather1: { id: 147, channel: 1, temperature: 9.4, humidity: 41, battery: 'Good' }
12:59:28debug [pimatic-homeduino]: received: [ 554, 1892, 3842, 9148 ] 01020102020101020101020202010101010101010102010202020201010102010201010203
12:59:28debug [pimatic-homeduino]: data: "RF receive 554 1892 3842 9148 0 0 0 0 01020102020101020101020202010101010101010102010202020201010102010201010203"
```

**Comment by elektronaut69** [comment]
The sensors are TFA and work in pilight. They did work until i flashed the nano in pimatic too.

**Comment by elektronaut69** [comment]
Found an old nano with some older homeduino sketch. I cannot tell you which version though because I do not know how to retrieve it. With this sketch I can receive and decode messages from the TFA sensors. So there is either something wrong with the nano I am using at the moment, or there is a problem with the sketch since the last update.

**Comment by sweetpi** [comment]
Could you post the raw timings (debug output) with the old nano?

**Comment by elektronaut69** [comment]
O.K. let's try to clarify this issue:

I have 3 nanos in total. All are chinese clones but one is from another vendor. With to of them I had to solder the pins myself. 

At the moment I use the prebuilt one. I didn't use it before, because D2 was not working. But I found out, that using the other receiver-pin its working well. The homeduino-sketch on this one is some weeks old. 

With the other two nanos receiving works not as well. Maybe I should improve my soldering skills ;-). But before flashing the last homeduino-build I was able to receive messages and pimatic did the decoding properly. After flashing one of those nanos pimatic did no longer decode the weather3 protocol properly. I did some testing yesterday and as I have a working config I did flash the second one with the latest build. After that the problem occured with this mcu too. As I said before, both of those nanos are from the same vendor, I had to solder the pins myself and receiving messages works not as well as with the nano that came prebuilt. So it is also possible, that those two nanos are flawed. 

Here are some raw timings from my working config:

```
debug [pimatic-homeduino]: weather3: { id: 123, channel: 1, temperature: 19, humidity: 41 }
15:06:42debug [pimatic-homeduino]: received: [ 679, 1879, 3771, 7580 ] 0101010202020201020201010101020102010101010201020201020101020101020101020101020202020303
15:06:42debug [pimatic-homeduino]: data: "RF receive 679 1879 3771 7580 0 0 0 0 0101010202020201020201010101020102010101010201020201020101020101020101020101020202020303"
14:58:56debug [pimatic-homeduino]: weather3: { id: 234, channel: 2, temperature: 21.7, humidity: 50 }
14:58:56debug [pimatic-homeduino]: received: [ 687, 1854, 3766, 7602 ] 0101020202010201020102010102020102010102010101020201010102010101020201010201020102020303
14:58:56debug [pimatic-homeduino]: data: "RF receive 687 1854 3766 7602 0 0 0 0 0101020202010201020102010102020102010102010101020201010102010101020201010201020102020303"
14:57:32debug [pimatic-homeduino]: weather3: { id: 123, channel: 1, temperature: 18.9, humidity: 41 }
14:57:32debug [pimatic-homeduino]: received: [ 676, 1876, 3780, 7580 ] 0101010202020201020201010101020101020101010201020201020101020101020101020101010201020303
14:57:32debug [pimatic-homeduino]: data: "RF receive 676 1876 3780 7580 0 0 0 0 0101010202020201020201010101020101020101010201020201020101020101020101020101010201020303"
```
I hope that helps.

**Comment by elektronaut69** [comment]
Borrowed an original arduino-nano to test. The problem is there with the latest sketch also. So it is not a problem with the nano clones. 

This is what I get. Protocol weather3.
```
23debug [pimatic-homeduino]: received: [ 703, 1850, 3749, 7600 ] 01010202020102010201010101020201010201020201010202010201020201010201010101010101010103
18:50:23debug [pimatic-homeduino]: data: "RF receive 703 1850 3749 7600 0 0 0 0 01010202020102010201010101020201010201020201010202010201020201010201010101010101010103"
18:50:23debug [pimatic-homeduino]: received: [ 683, 1866, 3778, 7632 ] 01010202020102010201010101020201010201020201010202010201020201010201010101010101010103
18:50:23debug [pimatic-homeduino]: data: "RF receive 683 1866 3778 7632 0 0 0 0 01010202020102010201010101020201010201020201010202010201020201010201010101010101010103"
```

**Comment by Icesory** [comment]
@sweetpi the implemented OneWire library disabled the global interrupts while it performs actions.
This is a big disadvantage about this library. I think the Onewire must be disable.

**Comment by sweetpi** [comment]
@Icesory Thanks for the hint. I'm just confused because the [pilight wiki says 86 pulses](http://wiki.pilight.org/doku.php/tfa_v6_0), so it looks like the new (not working) receivings are correct and the old ones are too long. But this would imply the protocol is wrong and we had an error in rfcontrol in the past.

I moved the onewire version of homeduino to a seperate branch some days ago. @elektronaut69 Could you please post a pulse sequence again with the most up to date master version?

**Comment by elektronaut69** [comment]
Here you are:
```
debug [pimatic-homeduino]: received: [ 555, 1888, 3846, 9156 ] 01020102020101020101020202010101010101010101010101020101010201020102020103
01:56:11debug [pimatic-homeduino]: data: "RF receive 555 1888 3846 9156 0 0 0 0 01020102020101020101020202010101010101010101010101020101010201020102020103"
01:56:07debug [pimatic-homeduino]: received: [ 670, 1836, 3730, 7632 ] 01010201020201020102010102010101010101010201010202010201010201010201010101010202010103
01:56:07debug [pimatic-homeduino]: data: "RF receive 670 1836 3730 7632 0 0 0 0 01010201020201020102010102010101010101010201010202010201010201010201010101010202010103"
01:56:07debug [pimatic-homeduino]: received: [ 657, 1848, 3749, 7580 ] 01010201020201020102010102010101010101010201010202010201010201010201010101010202010103
01:56:07debug [pimatic-homeduino]: data: "RF receive 657 1848 3749 7580 0 0 0 0 01010201020201020102010102010101010101010201010202010201010201010201010101010202010103"
```

**Comment by Icesory** [comment]
That is a really interesting bug. The protocol implementation in pimatic described two footer but the protocol has only one. 
@sweetpi do you have changed something int RFControl lib in the past? For me it looks like a different in the verification methods. 

But generally we have now a little problem. With future updates of pimatic-homeduino it is eventually necessary to update the Arduino too. I think it should not be hard to implement a update function for the Arduino in pimatic.

**Comment by elektronaut69** [comment]
I think I missed your update of yesterday. Here are some pulse sequences received with the latest update:
```
debug [pimatic-homeduino]: received: [ 701, 1841, 3748, 7584 ] 01010202020102010201010101020202010101020101010202010201010101010201010101010102020103
14:08:01debug [pimatic-homeduino]: data: "RF receive 701 1841 3748 7584 0 0 0 0 01010202020102010201010101020202010101020101010202010201010101010201010101010102020103"
14:08:01debug [pimatic-homeduino]: received: [ 679, 1864, 3768, 7600 ] 01010202020102010201010101020202010101020101010202010201010101010201010101010102020103
14:08:01debug [pimatic-homeduino]: data: "RF receive 679 1864 3768 7600 0 0 0 0 01010202020102010201010101020202010101020101010202010201010101010201010101010102020103"
14:07:51debug [pimatic-homeduino]: received: [ 671, 1837, 3736, 7568 ] 01010201020201020102010102010101020101010201010202010202020101010201010101010201010103
14:07:51debug [pimatic-homeduino]: data: "RF receive 671 1837 3736 7568 0 0 0 0 01010201020201020102010102010101020101010201010202010202020101010201010101010201010103"
14:07:51debug [pimatic-homeduino]: received: [ 652, 1854, 3757, 7592 ] 01010201020201020102010102010101020101010201010202010202020101010201010101010201010103
14:07:51debug [pimatic-homeduino]: data: "RF receive 652 1854 3757 7592 0 0 0 0 01010201020201020102010102010101020101010201010202010202020101010201010101010201010103"
```

**Comment by sweetpi** [comment]
Should be fixed if you update to pimatic-homeduino v0.8.44.

> That is a really interesting bug. The protocol implementation in pimatic described two footer but the protocol has only one. 
> @sweetpi do you have changed something int RFControl lib in the past? For me it looks like a different in the verification methods.

homeduino used your new verification method. The rfcontroljs submodule was accidentally linked to the development branch. I merged the new verification method to master and updated the weather3 protocol.

> But generally we have now a little problem. With future updates of pimatic-homeduino it is eventually necessary to update the Arduino too. I think it should not be hard to implement a update function for the Arduino in pimatic.

I think that's a good idea. At least we must have a version number that gets returned with the ready to detect which version is running and at least output a update notice.

**Comment by elektronaut69** [comment]
Receiving works now. But there is a new problem with sending. I will open a new issue for this.
